compaction/compaction.cc

```diff
@@ -1595,7 +1596,7 @@ class resharding_compaction final : public compaction {
     }
 
     compaction_writer create_compaction_writer(const dht::decorated_key& dk) override {
-        auto shard = dht::shard_of(*_schema, dk.token());
+        auto shard = _table_s.get_table().shard_of(dk.token());
         auto sst = _sstable_creator(shard);
```

Comment:
table_state abstracts away the concept of table. We need to add shard_of() method into table_state interface, rather than adding an accessor to underlying table. Unit tests for example might interact with compaction layer without actually having an actual table backing it.

Reply:
There's no table_state implementation which doesn't have a table right now.

What's the reason for the abstraction? We end up reflecting most of a table provides.

Reply:
we're abstracting the object being compacted with it. be it a table, or compaction group, or whatever. adding a table to table_state interface is a tremendous step back. we did lots of effort to remove direct table reference from the compaction layer. please add shard_of() to table_state instead, and whatever else you need.

Reply:
By the way, table_state is a terrible name. It's the compaction layer view of compaction groups.

Reply:
Every compaction object has a table, so it doesn't break the abstraction.I don't see how adding shard_of() is good, but adding table() is bad.

Reply:
That's a layer violation. Now you have table_state (compaction group) -> table -> all compaction groups (including the one being identified by table_state), and compaction group can be "casted" into table_state. So you're also adding a cyclic dependence.

I am even planning to rename table_state to compaction_group. compaction_group is only a group of sstables which are eligible to be compacted together. It doesn't have to have a table backing it.

And I want compaction tests in the future to work without an actual table.

So please just add shard_of() to table_state.

Reply:
Theoretically, not even shard_of() should belong to compaction_group. Compaction group should tell more things like if it owns a particular token. Which shard owns a particular token is indeed beyond its domain.

In perform_cleanup() interface in compaction manager, we have:

future<> perform_cleanup(owned_ranges_ptr sorted_owned_ranges, compaction::table_state& t);

So you can see that we provide sorted owned ranges in addition to the table state (group).

I can see that resharding is currently the sole user of shard_of(). So perhaps we can feed the function as a lambda param to resharding compaction.

Reply:
We can potentially feed table::shard_of() into resharding with compaction_descriptor. See how reshard() in distributed_loader.cc feeds a sstable creator lambda into compaction.

Reply:
> The kind of cycle you mention is already there: table -> compaction_group::table_state -> table.
> 
> table_state delegates to the table. Both implementations. This suggests that the layer separation at this level is artificial, in practice you want a table. In tests you also delegate to the table because otherwise you would have to reimplement the table. Abstracting this doesn't give us any benefit, just impediment for development.
> 

Memtables + sstables are now implemented in replica::compaction_group. So theoretically my test can implement table_state interface with c__g only, and not a replica::table. But I understand your point.

> > Theoretically, not even shard_of() should belong to compaction_group. So perhaps we can feed the function as a lambda param to resharding compaction.
> 
> Makes sense, I'll try that.

Thanks.

Reply:
table_state::get_table() doesn't exist in the current code.

A lot of table_state's methods return the same thing for all compaction groups, while others will return different things. This suggests that the actual state is in the second class of methods, while the first class describes configuration, not state.

```c++
// config
    virtual const schema_ptr& schema() const noexcept = 0;
    // min threshold as defined by table.
    virtual unsigned min_compaction_threshold() const noexcept = 0;
    virtual bool compaction_enforce_min_threshold() const noexcept = 0;
    virtual sstables::compaction_strategy& get_compaction_strategy() const noexcept = 0;
    virtual reader_permit make_compaction_reader_permit() const = 0;
    virtual sstables::sstables_manager& get_sstables_manager() noexcept = 0;
    virtual sstables::shared_sstable make_sstable() const = 0;
    virtual sstables::sstable_writer_config configure_writer(sstring origin) const = 0;
    virtual bool is_auto_compaction_disabled_by_user() const noexcept = 0;
    virtual bool tombstone_gc_enabled() const noexcept = 0;
// state
    virtual const sstables::sstable_set& main_sstable_set() const = 0;
    virtual const sstables::sstable_set& maintenance_sstable_set() const = 0;
    virtual std::unordered_set<sstables::shared_sstable> fully_expired_sstables(const std::vector<sstables::shared_sstable>& sstables, gc_clock::time_point compaction_time) const = 0;
    virtual const std::vector<sstables::shared_sstable>& compacted_undeleted_sstables() const noexcept = 0;
    virtual compaction_strategy_state& get_compaction_strategy_state() noexcept = 0;
    virtual api::timestamp_type min_memtable_timestamp() const = 0;
    virtual future<> on_compaction_completion(sstables::compaction_completion_desc desc, sstables::offstrategy offstrategy) = 0;
    virtual const tombstone_gc_state& get_tombstone_gc_state() const noexcept = 0;
    virtual compaction_backlog_tracker& get_backlog_tracker() = 0;
    virtual const std::string& get_group_id() const noexcept = 0;
    virtual seastar::condition_variable& get_staging_done_condition() noexcept = 0; // ?
```